csrc/preseg_passes/insert_reshardings.cpp

```diff
@@ -129,6 +130,56 @@ void insertReshardingSetsAfter(Fusion* fusion) {
   }
 }
 
+// Canonicalizes tv's loop domain for simplicity and working around schedulers'
+// limitations. Many schedulers panic when they see the input fusion segment
+// contains non-DID loop splits. For example, an rFactor tensor may look like
+// the following:
+//
+//                            r{k}
+//                            /  \.
+// [i{m}         i{n}    iDIDx{d}  r{k/d}]
+//               /  \.
+//            i{d} i{n/d}
+//
+// The split of i{n} is unnecessary because i{d} and i{n/d} are both
+// ParallelType::Serial. This function replaces the two with i{n} in the loop
+// domain.
+void canonicalizeLoopDomain(TensorView* tv) {
+  LinkedHashMap<IterDomain*, std::monostate> loop;
+  for (IterDomain* id : tv->getLoopDomain()) {
+    loop.pushBack(id, std::monostate());
+  }
+
+  for (Expr* transform :
+       DependencyCheck::getAllExprsBetween(
+           {tv->getLogicalDomain().begin(), tv->getLogicalDomain().end()},
+           {tv->getLoopDomain().begin(), tv->getLoopDomain().end()}) |
+           std::views::reverse) {
+    auto* split = dynamic_cast<Split*>(transform);
+    if (split == nullptr) {
+      continue;
+    }
+
+    if (split->outer()->isParallelized() || split->inner()->isParallelized()) {
+      continue;
+    }
+
+    if (!loop.contains(split->outer()) || !loop.contains(split->inner())) {
+      continue;
```

Comment:
Is this situation possible? Can you give me an example?

Reply:
```
    a
  /   \
 b     c
      / \
DIDx{d}  e
```
Two splits. Handling the split on `c` won't change loop because `d` is parallelized, so `loop` remains `[b, d, e]`. Then, handling the split on `a` won't see `c`.